src/bun.js/node/node_fs.zig

```diff
@@ -242,111 +242,491 @@ pub const Async = struct {
+        /// When clonefile cannot be used, this task is started once per file.
+        pub const SingleTask = struct {
+            cp_task: *ThisAsyncCpTask,
+            src: bun.OSPathSliceZ,
+            dest: bun.OSPathSliceZ,
+            task: JSC.WorkPoolTask = .{ .callback = &SingleTask.workPoolCallback },
 
-        var node_fs = NodeFS{};
-        node_fs.cpAsync(this);
-    }
+            const ThisSingleTask = @This();
 
-    /// May be called from any thread (the subtasks)
-    fn finishConcurrently(this: *AsyncCpTask, result: Maybe(Return.Cp)) void {
-        if (this.has_result.cmpxchgStrong(false, true, .Monotonic, .Monotonic)) |_| {
-            return;
+            pub fn create(
+                parent: *ThisAsyncCpTask,
+                src: bun.OSPathSliceZ,
+                dest: bun.OSPathSliceZ,
+            ) void {
+                var task = bun.new(ThisSingleTask, .{
+                    .cp_task = parent,
+                    .src = src,
+                    .dest = dest,
+                });
+
+                JSC.WorkPool.schedule(&task.task);
+            }
+
+            fn workPoolCallback(task: *JSC.WorkPoolTask) void {
+                var this: *ThisSingleTask = @fieldParentPtr(ThisSingleTask, "task", task);
+
+                // TODO: error strings on node_fs will die
+                var node_fs = NodeFS{};
+
+                const args = this.cp_task.args;
+                const result = node_fs._copySingleFileSync(
+                    this.src,
+                    this.dest,
+                    @enumFromInt((if (args.flags.errorOnExist or !args.flags.force) Constants.COPYFILE_EXCL else @as(u8, 0))),
+                    null,
+                    this.cp_task.args,
+                );
+
+                brk: {
+                    switch (result) {
+                        .err => |err| {
+                            if (err.errno == @intFromEnum(E.EXIST) and !args.flags.errorOnExist) {
+                                break :brk;
+                            }
+                            this.cp_task.finishConcurrently(result);
+                            this.deinit();
+                            return;
+                        },
+                        .result => {
+                            this.cp_task.onCopy(this.src, this.dest);
+                        },
+                    }
+                }
+
+                const old_count = this.cp_task.subtask_count.fetchSub(1, .Monotonic);
+                if (old_count == 1) {
+                    this.cp_task.finishConcurrently(Maybe(Return.Cp).success);
+                }
+
+                this.deinit();
+            }
+
+            pub fn deinit(this: *ThisSingleTask) void {
+                // There is only one path buffer for both paths. 2 extra bytes are the nulls at the end of each
+                bun.default_allocator.free(this.src.ptr[0 .. this.src.len + this.dest.len + 2]);
+
+                bun.destroy(this);
```

Comment:
not sure why this is using `bun.destroy` / `bun.new` instead of `usingnamespace bun.New` and ThisSingleTask.new / this.destroy